File: backend/app/api/repository.py
import time

from fastapi import APIRouter

from app.core.logger import logger
from app.schemas.repository import (
    RepositoryImportRequest,
    RepositoryMetadata,
    RepositorySummaryResponse,
)
from app.services.repository_analyzer import RepositoryAnalyzer
from app.services.repository_file_reader_service import RepositoryFileReaderService
from app.services.repository_knowledge_builder import RepositoryKnowledgeBuilder
from app.services.repository_summary_service import RepositorySummaryService

router = APIRouter(prefix="/api/repositories", tags=["repositories"])

repository_analyzer = RepositoryAnalyzer()
repository_file_reader_service = RepositoryFileReaderService()
repository_knowledge_builder = RepositoryKnowledgeBuilder()
repository_summary_service = RepositorySummaryService()


@router.post("/import", response_model=RepositoryMetadata)
async def import_repository(request: RepositoryImportRequest) -> RepositoryMetadata:
    """Fetch metadata for a GitHub repository from the provided URL."""
    return await repository_analyzer.get_metadata(request.url)


@router.post("/summary", response_model=RepositorySummaryResponse)
async def summarize_repository(
    request: RepositoryImportRequest,
) -> RepositorySummaryResponse:
    """Run the full analysis pipeline and generate an AI repository summary."""
    pipeline_start = time.perf_counter()
    logger.info("Summary Pipeline Started")
    logger.info("Repository: %s", request.url)

    analysis = await repository_analyzer.analyze(request.url)
    documents = repository_file_reader_service.collect_documents(analysis.local_path)
    knowledge = repository_knowledge_builder.build(analysis, documents)
    summary = await repository_summary_service.generate_summary(knowledge)

    elapsed = time.perf_counter() - pipeline_start
    logger.info("Summary Pipeline Finished")
    logger.info("Total Time: %.2fs", elapsed)

    return RepositorySummaryResponse(
        summary=summary,
        metadata=analysis.metadata,
        tree=analysis.tree,
    )

File: frontend/src/services/repository.ts
import type {
  RepositoryMetadata,
  RepositorySummaryResponse,
} from "@/types/repository";

const API_BASE = "http://127.0.0.1:8000";

export async function importRepository(
  url: string,
): Promise<RepositoryMetadata> {
  const response = await fetch(`${API_BASE}/api/repositories/import`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ url }),
  });

  if (!response.ok) {
    const body = await response.json().catch(() => null);
    const detail = body?.detail;
    const message =
      typeof detail === "string"
        ? detail
        : `Request failed with status ${response.status}`;
    throw new Error(message);
  }

  return response.json();
}

export async function summarizeRepository(
  url: string,
): Promise<RepositorySummaryResponse> {
  const response = await fetch(`${API_BASE}/api/repositories/summary`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ url }),
  });

  if (!response.ok) {
    const body = await response.json().catch(() => null);
    const detail = body?.detail;
    const message =
      typeof detail === "string"
        ? detail
        : `Request failed with status ${response.status}`;
    throw new Error(message);
  }

  return response.json();
}
